Reduce llvm-gsymutil memory usage

kevinfrei opened the pull request:
llvm-gsymutil eats a lot of RAM. On some large binaries, it causes OOM's on smaller hardware, consuming well over 64GB of RAM. This change frees line tables once we're done with them, and frees DWARFUnits's DIE's when we finish processing each DU, though they may get reconstituted if there are references from other DU's during processing. Once the conversion is complete, all DIE's are freed. The reduction in peak memory usage from these changes showed between 7-12% in my tests.

The double-checked locking around the creation & freeing of the data structures was tested on a 166 core system. I validated that it trivially malfunctioned without the locks (and with stupid reordering of the locks) and worked reliably with them.

Commit:
Release line tables when we're done with them

Commit:
Finished without crashing

Commit:
Added locking

Commit:
Code formatting

kevinfrei commented:
@clayborg @jeffreytan81 could one of you please take a look?

jeffreytan81 reviewed (commented):
Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?

clayborg commented on llvm/include/llvm/DebugInfo/DWARF/DWARFUnit.h:
```diff
@@ -257,6 +258,9 @@ class DWARFUnit {
 
   std::shared_ptr<DWARFUnit> DWO;
 
+  mutable llvm::sys::RWMutex m_cu_die_array_mutex;
```
LLDB coding guidelines are a bit different for instance variables: to don't prefix with `m_` and they use camel case. So this should be something like:
```
mutable llvm::sys::RWMutex  CUDieArrayMutex;
```

clayborg commented on llvm/include/llvm/DebugInfo/DWARF/DWARFUnit.h:
```diff
@@ -257,6 +258,9 @@ class DWARFUnit {
   std::shared_ptr<DWARFUnit> DWO;
 
+  mutable llvm::sys::RWMutex m_cu_die_array_mutex;
+  mutable llvm::sys::RWMutex m_all_die_array_mutex;
```
`AllDieArrayMutex`

clayborg commented on llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:
```diff
@@ -496,13 +496,29 @@ void DWARFUnit::extractDIEsIfNeeded(bool CUDieOnly) {
+    if (DieArray.empty())
+      return Error::success();
+
+    // If CU DIE was just parsed, copy several attribute values from it.
```
This comment doesn't match what is happening. Is this from a copy/paste?

kevinfrei commented on llvm/include/llvm/DebugInfo/DWARF/DWARFUnit.h:
```diff
@@ -257,6 +258,9 @@ class DWARFUnit {
 
   std::shared_ptr<DWARFUnit> DWO;
 
+  mutable llvm::sys::RWMutex m_cu_die_array_mutex;
```
Gack. Jumping between LLVM and LLDB conventions. Sorry. I'll fix this.

kevinfrei commented:
> Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?

They're there for protected the allocation & population, or destruction of the DieArrays, not generally reading/writeing those fields. There needs to be two (for just the first CUDie, and then the whole array) as populating the entire array can trigger a request to populate just the first.

Commit:
Updated from code review feedback

clayborg commented on llvm/lib/DebugInfo/GSYM/DwarfTransformer.cpp:
```diff
@@ -612,11 +617,16 @@ Error DwarfTransformer::convert(uint32_t NumThreads, OutputAggregator &Out) {
       if (Die) {
         CUInfo CUI(DICtx, dyn_cast<DWARFCompileUnit>(CU.get()));
-        pool.async([this, CUI, &LogMutex, &Out, Die]() mutable {
+        pool.async([this, CUI, &CU, &LogMutex, &Out, Die]() mutable {
```
If "CU" is a pointer, do we need to capture a reference? Could this just be:
```
pool.async([this, CUI, CU, &LogMutex, &Out, Die]() mutable {
```
Or does capture make `CU` into a `const *`?

kevinfrei replied:
Not a pointer: a unique_ptr, so only capture by reference.

clayborg reviewed (commented):
I am fine with the GSYM parts. I will let the LLVM DWARF experts accept the changes in `DWARFUnit.h/.cpp`

ayermolo commented:
> > Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?
> 
> They're there for protecting the allocation+population, or destruction of the DieArrays, not generally reading/writing those fields. There needs to be two (for just the first CUDie, and then the whole array) as populating the first can trigger a request to populate the rest.

What is the code path for that?

kevinfrei commented:
> > > Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?
> > 
> > 
> > They're there for protecting the allocation+population, or destruction of the DieArrays, not generally reading/writing those fields. There needs to be two (for just the first CUDie, and then the whole array) as populating the first can trigger a request to populate the rest.
> 
> What is the code path for that?

Sorry: Wrote that backwards. Populating the whole array can trigger demand population of the UnitDIE through the function `determineStringOffsetsTableContribution`. 

> > > Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?
> > 
> > 
> > They're there for protecting the allocation+population, or destruction of the DieArrays, not generally reading/writing those fields. There needs to be two (for just the first CUDie, and then the whole array) as populating the first can trigger a request to populate the rest.
> 
> What is the code path for that?

Sorry: I wrote that backwards (fixed the original comment). Populated the full array can trigger a request to populate the CuDie through `determineStringOffsetsTableContribution`, which is why you need the two different locks. The first is always created before the list, but you need to lock it separately so that the read-lock will be free to return just the CUDie.

ayermolo commented:
> > > > Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?
> > > 
> > > 
> > > They're there for protecting the allocation+population, or destruction of the DieArrays, not generally reading/writing those fields. There needs to be two (for just the first CUDie, and then the whole array) as populating the first can trigger a request to populate the rest.
> > 
> > 
> > What is the code path for that?
> 
> Sorry: I wrote that backwards (fixed the original comment). Populating the full array can trigger a request to populate the CuDie through `determineStringOffsetsTableContribution`, which is why you need the two different locks. The first is always created before the list, but you need to lock it separately so that the read-lock will be free to return just the CUDie.

Glanced at the code. I think I am missing something. How can code through determineStringOffsetsTableContribution trigger the path? It just accesses StrOffset Section.

kevinfrei commented:
> > > > > Per my understanding, these two mutex are required to protect any read/write to DieArray field? If so, there seem to be many more accesses of this field not protected?
> > > > 
> > > > 
> > > > They're there for protecting the allocation+population, or destruction of the DieArrays, not generally reading/writing those fields. There needs to be two (for just the first CUDie, and then the whole array) as populating the first can trigger a request to populate the rest.
> > > 
> > > 
> > > What is the code path for that?
> > 
> > 
> > Sorry: I wrote that backwards (fixed the original comment). Populating the full array can trigger a request to populate the CuDie through `determineStringOffsetsTableContribution`, which is why you need the two different locks. The first is always created before the list, but you need to lock it separately so that the read-lock will be free to return just the CUDie.
> 
> Glanced at the code. I think I am missing something. How can code through determineStringOffsetsTableContribution trigger the path? It just accesses StrOffset Section.

Yeah, I didn't see it by looking at the code, because the recursion occurs inside getUnitDIE() (in the .h file), which calls extractDIEsIfNeeded(...)

kevinfrei commented:
ping @ayermolo @dwblaikie This is blocking some dwarf-health tracking work I'm trying to make headway on. The only 'stylistic' adjustment I can see making sense is to have some sort of doubled-checked locking abstraction that takes a lock and a couple lambda's to make the duplicated code cleaner, which I'm happy to do if someone wants it, but I haven't seen something like that elsewhere.

ayermolo commented:
OK. I think this two lock thing makes sense in my head.
One concern I have is that between locks in tryExtractDIEsIfNeeded it switches to clearDIEs.

Commit:
Using a helper to make the double checked locking more legible

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D57863981

Commit:
Added the top level lock to prevent clearDies from causing problems

kevinfrei commented:
@ayermolo I used a rwlock model to prevent the issue you identified. Good catch!

kevinfrei commented:
To alleviate copy-pasta problems, I've added a local DoubleCheckewhatdRWLocker helper function that takes the "check if it's already been initialized" lambda, and the initializer lambda (and the lock). I feel like the result is a much cleaner version of the original diff, plus it has the third lock to prevent free's from occurring between the single CU and the array CU initialization code that @ayermolo saw. @dwblaikie they tell me you're the guy who needs to approve this, now...

dwblaikie commented:
Ah, I thought this was all in gsym, so hadn't given the review much attention.

Are these independent changes - the line table clearing stuff presumably is independent, for instance? I guess the unit DIE  clearing is dependent on the extra locking being added?

The extra locking is unfortunate/more difficult to just straight up approve, as we'd previously restricted the multithreading support to the DWARFContext by having distinct contexts - so that non-multithreaded code didn't take the hit or complexity for the multithreaded code. But this new codepath and locking is present for single and multithreaded uses alike, by the looks of it? Not sure what to do about that...

kevinfrei commented:
> Are these independent changes - the line table clearing stuff presumably is independent, for instance? I guess the unit DIE clearing is dependent on the extra locking being added?

I'm only doing one "real" thing here, so no, the changes can't be split into smaller, still useful/sensible pieces (except that I could delay the gsymutil change until after the dwarfunit support for releasing the memory is in, which doesn't seem worthwhile)

> The extra locking is unfortunate/more difficult to just straight up approve, as we'd previously restricted the multithreading support to the DWARFContext by having distinct contexts - so that non-multithreaded code didn't take the hit or complexity for the multithreaded code. But this new codepath and locking is present for single and multithreaded uses alike, by the looks of it? Not sure what to do about that...

I spent about 3 days staring at how to accomplish this in a way that wouldn't increase the cost of single-threaded code, and couldn't come up with anything other that copying the entire class and making a multi-threaded 'freeable' duplicate, which seemed like an awful lot of code duplication to save a few hundred CPU cycles in code that is *egregiously* I/O bound. (I have a 166 core server: gsymutil works fastest when it's run on 6-10 cores: any more swamps the NVMe drive).

kevinfrei commented:
All that said, apparently something broke a bunch of tests (I hadn't run them since my last couple changes) so I'm not blocked on your review just yet :/

Commit:
Fixed the error handling issues; fully tested

kevinfrei commented:
@dwblaikie If you could take a look, now, I'd appreciate it. Using move semantics to force error handling makes for extra fun with lambda's :)

kevinfrei commented on llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:
```diff
@@ -495,21 +495,73 @@ void DWARFUnit::extractDIEsIfNeeded(bool CUDieOnly) {
+    return ((CUDieOnly && !DieArray.empty()) || DieArray.size() > 1);
+  };
+
+  // Lambda to extract the CU DIE.
```
I showed the code to a coworker with lots more recent experience in LLVM, and he suggested inlining the lambdas (which was nearly unreadable originally), but breaking the code into pieces so you can see the locking at each point (which makes the code more legible). I'm gonna get that change up hopefully by end of day, then you can tell me which you prefer.

ayermolo replied:
I vote for readable. :D

kevinfrei replied:
> I vote for readable. :D
> 
> I think I remember your internal one, and this one is more readable. at least to me.

Yeah, this is better than the previous iteration, but broken up in even smaller pieces looks downright well thought out. Update forthcoming...

Commit:
Refactored to make the overall code read more cleanly

kevinfrei commented on llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:
```diff
@@ -658,6 +711,9 @@ void DWARFUnit::clearDIEs(bool KeepCUDie) {
   // Create a new vector with a small capacity and assign it to the DieArray to
   // have previous contents freed.
+  llvm::sys::ScopedWriter FreeLock(CUDieFreeMutex);
+  llvm::sys::ScopedWriter CULock(CUDieArrayMutex);
```
I *think* taking all 3 locks is overkill; If you ahve CUDieFreeMutex, the other two are guaranteed to be available.

kevinfrei commented on llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:
```diff
@@ -658,6 +711,9 @@ void DWARFUnit::clearDIEs(bool KeepCUDie) {
   // It depends on the implementation whether the request is fulfilled.
   // Create a new vector with a small capacity and assign it to the DieArray to
   // have previous contents freed.
+  llvm::sys::ScopedWriter FreeLock(CUDieFreeMutex);
+  llvm::sys::ScopedWriter CULock(CUDieArrayMutex);
+  llvm::sys::ScopedWriter AllLock(AllDieArrayMutex);
```
I believe taking all 3 locks is overkill, as getting the CUDieFreeMutex writer lock will block until the CUDie and AllDie locks are free. Thoughts?

dwblaikie replied:
Yeah, sounds about right.

And perhaps once the other two are removed the code'll be basically the same as it was without this patch (except for two lock acquisitions) before and so a lot of the concerns/questions/issues (the newly subdivided functions and names are quite confusing to me, but I've deleted my comments on that so as not to confuse things further if it's all going away anyway) go away?

kevinfrei replied:
I can't remove the other two locks. They have to stay, as the freeing mechanism requires it. The third lock (Free) is specifically to deal with the issue that @ayermolo pointed out a couple revisions ago. (Freeing could free just the CU DIE while the non-CU DIE stuff was being initialized, resulting in an invalid state of no CU DIE, but non-CU DIE's are parsed). The entire process is kind of a mess, but I need a "CU DIE initialization" RWLock, plus the "Non-CU DIE initialization" lock as two separate entities, as initializing the non-CU DIE's can trigger recursion (you have to be able to grab the reader lock of the CU DIE init to tell that it's already been initialized, then bail)

kevinfrei replied:
@dwblaikie what are you thoughts here? I've updated the diff to use the 3 locks as efficiently as I can come up with. If you'd like to chat in real time, I'm happy to hop on a VC in whatever host system you're comfortable with (and my calendar is fully blocked off all day Friday) I really do need this diff landed.

dwblaikie replied:
Sorry for the delays & thanks for the offer to meet more directly (& setting aside time - I didn't see this message until later, that week was a bit busy for me I think... ) - looking into this now.

Commit:
Removed the acquisition of extra, unecessary locks

kevinfrei commented:
@dwblaikie friendly ping...

Commit:
Renamed locks and helper methods, improved comments

kevinfrei commented:
@dwblaikie ping

dwblaikie commented:
Is it worth having 3 separate mutexes? One answer would be to go fully coarse grained - just have 1 that covers the whole array. Does that have noticably worse performance for your use case?

The other option would be 2 - you have one covering the first element of the array, and another covering the rest - I guess then you have problems when reallocating the array, though?

dwblaikie commented:
> > Is it worth having 3 separate mutexes? One answer would be to go fully coarse grained - just have 1 that covers the whole array. Does that have noticably worse performance for your use case?
> 
> I started with what you're describing, and it deadlocks, as there's a path from the non-CU DIE creation that eventually recurses into the function (which I called out in the [review above](https://github.com/llvm/llvm-project/pull/91023#issuecomment-2116631029))

Thanks, sorry I'm rehashing things. A recursive mutex would apply here, then? (& we even have this thing: https://llvm.org/doxygen/Mutex_8h_source.html - which would reduce the overhead (presumably?) to the non-multithreaded users (well, I guess only to users building all of LLVM without multithreading, not users building with multithreading enabled and just not using multiple threads to interact with libDebugInfoDWARF)?)

Does a single recursive mutex substantially affect the performance of the situations you're interested in?

kevinfrei commented:
> Does a single recursive mutex substantially affect the performance of the situations you're interested in?

Yes. You need a reader-writer mutex, as the reader scenario is *very* common, but conflicts enough to matter quite a bit. Any other performance doesn't matter at all. Even in single-threaded mode, the locking overhead isn't likely to register, as the work is *deeply* IO bound (and the existing lock is already one of the LLVM smart locks so it shouldn't be a cost)

dwblaikie reviewed (approved):
Fair enough - thanks for your patience walking me through it

kevinfrei commented on llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:
```diff
@@ -495,21 +495,78 @@ void DWARFUnit::extractDIEsIfNeeded(bool CUDieOnly) {
+// Helper to safely check if the Compile-Unit DIE has been extracted already.
+// If not, then extract it, and return false, indicating that it was *not*
+// already extracted.
+bool DWARFUnit::extractCUDieIfNeeded(bool CUDieOnly, bool &HasCUDie) {
```
Yup, you're correct: This is allocating (and extracting) all the DIE's. What's protected by the second lock appears to be extracting string (offset...), range, and location tables for the DIEs that were just extracted. Would renaming things to `extractDIEsIfNeeded`/`extractTablesForDIEsIfNeeded` seem reasonable?

(I'm also going to see if I can free those tables in a later diff, as I'd love to drive memory usage lower than it currently is even with this change)

kamaub commented:
It seems this commit caused a build failure on the [clang-ppc64le-linux-multistage](https://lab.llvm.org/buildbot/#/builders/76) build no.[583](https://lab.llvm.org/buildbot/#/builders/76/builds/583).

Please take a look at it and post a fix or revert your patch if the fix will take to long to implement. Let me know if there is anyway help or extra information I can offer.
```
FAILED: lib/libLLVMDebugInfoDWARF.so.19.0git 
: && /usr/lib64/ccache/c++ -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete   -Wl,-rpath-link,/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-multistage-test/clang-ppc64le-multistage/stage1/./lib  -Wl,--gc-sections -shared -Wl,-soname,libLLVMDebugInfoDWARF.so.19.0git -o lib/libLLVMDebugInfoDWARF.so.19.0git lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFAbbreviationDeclaration.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFAddressRange.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFAcceleratorTable.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFCompileUnit.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFContext.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDataExtractor.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugAbbrev.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugAddr.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugArangeSet.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugAranges.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugFrame.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFTypePrinter.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugInfoEntry.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugLine.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugLoc.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugMacro.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugPubTable.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugRangeList.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugRnglists.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDie.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFExpression.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFFormValue.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFGdbIndex.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFListTable.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFLocationExpression.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFTypeUnit.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnitIndex.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFVerifier.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib:/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-multistage-test/clang-ppc64le-multistage/stage1/lib:"  lib/libLLVMObject.so.19.0git  lib/libLLVMBinaryFormat.so.19.0git  lib/libLLVMTargetParser.so.19.0git  lib/libLLVMSupport.so.19.0git  -Wl,-rpath-link,/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-multistage-test/clang-ppc64le-multistage/stage1/lib && :
lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o: In function `DoubleCheckedRWLocker(llvm::sys::SmartRWMutex<false>&, std::function<bool ()> const&, std::function<void ()> const&)':
DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0x44): undefined reference to `pthread_rwlock_rdlock'
DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0x94): undefined reference to `pthread_rwlock_unlock'
DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0xa8): undefined reference to `pthread_rwlock_wrlock'
DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0x114): undefined reference to `pthread_rwlock_unlock'
lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o: In function `llvm::DWARFUnit::tryExtractDIEsIfNeeded(bool) [clone .localalias.18]':
DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit22tryExtractDIEsIfNeededEb+0x44): undefined reference to `pthread_rwlock_rdlock'
DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit22tryExtractDIEsIfNeededEb+0x94): undefined reference to `pthread_rwlock_unlock'
lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o: In function `llvm::DWARFUnit::clearDIEs(bool) [clone .localalias.15]':
DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit9clearDIEsEb+0x30): undefined reference to `pthread_rwlock_wrlock'
DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit9clearDIEsEb+0x7c): undefined reference to `pthread_rwlock_unlock'
collect2: error: ld returned 1 exit status
```

kevinfrei commented:
> It seems this commit caused a build failure on the [clang-ppc64le-linux-multistage](https://lab.llvm.org/buildbot/#/builders/76) build no.[583](https://lab.llvm.org/buildbot/#/builders/76/builds/583).
> 
> Please take a look at it and post a fix or revert your patch if the fix will take to long to implement. Let me know if there is anyway help or extra information I can offer.
> 
> ```
> FAILED: lib/libLLVMDebugInfoDWARF.so.19.0git 
> : && /usr/lib64/ccache/c++ -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete   -Wl,-rpath-link,/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-multistage-test/clang-ppc64le-multistage/stage1/./lib  -Wl,--gc-sections -shared -Wl,-soname,libLLVMDebugInfoDWARF.so.19.0git -o lib/libLLVMDebugInfoDWARF.so.19.0git lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFAbbreviationDeclaration.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFAddressRange.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFAcceleratorTable.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFCompileUnit.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFContext.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDataExtractor.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugAbbrev.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugAddr.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugArangeSet.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugAranges.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugFrame.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFTypePrinter.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugInfoEntry.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugLine.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugLoc.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugMacro.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugPubTable.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugRangeList.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDebugRnglists.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFDie.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFExpression.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFFormValue.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFGdbIndex.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFListTable.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFLocationExpression.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFTypeUnit.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnitIndex.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFVerifier.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib:/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-multistage-test/clang-ppc64le-multistage/stage1/lib:"  lib/libLLVMObject.so.19.0git  lib/libLLVMBinaryFormat.so.19.0git  lib/libLLVMTargetParser.so.19.0git  lib/libLLVMSupport.so.19.0git  -Wl,-rpath-link,/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-multistage-test/clang-ppc64le-multistage/stage1/lib && :
> lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o: In function `DoubleCheckedRWLocker(llvm::sys::SmartRWMutex<false>&, std::function<bool ()> const&, std::function<void ()> const&)':
> DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0x44): undefined reference to `pthread_rwlock_rdlock'
> DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0x94): undefined reference to `pthread_rwlock_unlock'
> DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0xa8): undefined reference to `pthread_rwlock_wrlock'
> DWARFUnit.cpp:(.text._ZL21DoubleCheckedRWLockerRN4llvm3sys12SmartRWMutexILb0EEERKSt8functionIFbvEERKS4_IFvvEE+0x114): undefined reference to `pthread_rwlock_unlock'
> lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o: In function `llvm::DWARFUnit::tryExtractDIEsIfNeeded(bool) [clone .localalias.18]':
> DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit22tryExtractDIEsIfNeededEb+0x44): undefined reference to `pthread_rwlock_rdlock'
> DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit22tryExtractDIEsIfNeededEb+0x94): undefined reference to `pthread_rwlock_unlock'
> lib/DebugInfo/DWARF/CMakeFiles/LLVMDebugInfoDWARF.dir/DWARFUnit.cpp.o: In function `llvm::DWARFUnit::clearDIEs(bool) [clone .localalias.15]':
> DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit9clearDIEsEb+0x30): undefined reference to `pthread_rwlock_wrlock'
> DWARFUnit.cpp:(.text._ZN4llvm9DWARFUnit9clearDIEsEb+0x7c): undefined reference to `pthread_rwlock_unlock'
> collect2: error: ld returned 1 exit status
> ```

Please revert it. Looks like the PPC64le build doesn't like lambda's very much. I'll try to refactor it without the lambda's, but I won't be able to do it in the next few hours :/
